src/adversarial/neurify.jl

```diff
@@ -0,0 +1,306 @@
+    UG = copy(max_violation_con)
+
+    i_max, j_max, influence_max = 0, 0, -Inf
+
+    for i in reverse(1:length(nnet.layers))
+        layer = nnet.layers[i]
+        if layer.activation isa ReLU
+            for j in 1:n_nodes(layer)
+                if is_ambiguous_activation(i, j)
+                    # taking `influence = max_gradient * reach.r[i][j]*k` would be
+                    # different from original paper, but can improve the split efficiency.
+                    # where `k = n-i+1`, i.e. counts up from 1 as you go back in layers.
+                    influence = max(abs(LG[j]), abs(UG[j])) * radii[i][j]
+                    if influence >= influence_max && (i, j, influence) ∉ splits
+                        i_max, j_max, influence_max = i, j, influence
+                    end
+                end
+            end
+        end
+
+        LG_hat = max.(LG, 0.0) .* LΛ[i] .+ min.(LG, 0.0) .* UΛ[i]
+        UG_hat = min.(UG, 0.0) .* LΛ[i] .+ max.(UG, 0.0) .* UΛ[i]
+
+        LG, UG = interval_map(layer.weights', LG_hat, UG_hat)
+    end
+
+    # NOTE can omit this line in the paper version
+    i_max == 0 || j_max == 0 && error("Can not find valid node to split")
+
+    push!(splits, (i_max, j_max, influence_max))
+
+    return (i_max, j_max, influence_max)
+end
+
+function forward_layer(solver::Neurify, layer::Layer, input)
+    return forward_act(forward_linear(solver, input, layer), layer)
+end
+
+# Symbolic forward_linear for the first layer
+function forward_linear(solver::Neurify, input::AbstractPolytope, layer::Layer)
+    (W, b) = (layer.weights, layer.bias)
+    sym = SymbolicInterval(hcat(W, b), hcat(W, b), input)
+    LΛ = Vector{Vector{Int64}}(undef, 0)
+    UΛ = Vector{Vector{Int64}}(undef, 0)
+    r = Vector{Vector{Int64}}(undef, 0)
+    return SymbolicIntervalGradient(sym, LΛ, UΛ, r)
+end
+
+# Symbolic forward_linear
+function forward_linear(solver::Neurify, input::SymbolicIntervalGradient, layer::Layer)
+    (W, b) = (layer.weights, layer.bias)
+    output_Up = max.(W, 0) * input.sym.Up + min.(W, 0) * input.sym.Low
+    output_Low = max.(W, 0) * input.sym.Low + min.(W, 0) * input.sym.Up
+    output_Up[:, end] += b
+    output_Low[:, end] += b
+    sym = SymbolicInterval(output_Low, output_Up, input.sym.interval)
+    return SymbolicIntervalGradient(sym, input.LΛ, input.UΛ, input.r)
+end
+
+# Symbolic forward_act
+function forward_act(input::SymbolicIntervalGradient, layer::Layer{ReLU})
+    n_node, n_input = size(input.sym.Up)
+    output_Low, output_Up = copy(input.sym.Low), copy(input.sym.Up)
+    mask_lower, mask_upper = zeros(Float64, n_node), ones(Float64, n_node)
+    interval_width = zeros(Float64, n_node)
+    for i in 1:n_node
+        # Symbolic linear relaxation
+        # This is different from ReluVal
+
+        up_low, up_up = bounds(input.sym.Up[i, :], input.sym.interval)
+        low_low, low_up = bounds(input.sym.Low[i, :], input.sym.interval)
+
+        interval_width[i] = up_up - low_low
```

Comment:
@changliuliu you mentioned we can drop the `r` field being made here since this should be calculable using the other fields where we need it in `get_max_nodewise_influence`. How would we calculate it there?

Reply:
We just need to do the same calculation here. Get the symbolic representation `sym`, then compute `up_up` and `low_low`. Then `r` is the difference between the two.

Reply:
Looks like where we need `r` we've actually lost most of the info needed to easily get it (in the backprop in nodewise_influence). What we have there is the symbolic interval at the output, and we need the symbolic layer at each layer `i` to get its radius. That's a bit annoying...